src/pip/_internal/vcs/git.py

```diff
@@ -74,27 +74,41 @@ def export(self, location):
                 show_stdout=False, cwd=temp_dir.path
             )
 
-    def check_rev_options(self, rev, dest, rev_options):
-        """Check the revision options before checkout to compensate that tags
-        and branches may need origin/ as a prefix.
-        Returns the SHA1 of the branch or tag if found.
+    def get_revision_sha(self, dest, rev):
         """
-        revisions = self.get_short_refs(dest)
-
-        origin_rev = 'origin/%s' % rev
-        if origin_rev in revisions:
-            # remote branch
-            return [revisions[origin_rev]]
-        elif rev in revisions:
-            # a local tag or branch name
-            return [revisions[rev]]
-        else:
+        Return a commit hash for the given revision if it names a remote
+        branch or tag.  Otherwise, return None.
+
+        Args:
+          dest: the repository directory.
+          rev: the revision name.
+        """
+        # Pass rev to pre-filter the list.
+        output = self.run_command(['show-ref', rev], cwd=dest,
+                                  show_stdout=False, on_returncode='ignore')
+        refs = {}
+        for line in output.strip().splitlines():
+            sha, ref = line.split(None, 1)
+            refs[ref] = sha
+
+        sha = refs.get('refs/remotes/origin/{}'.format(rev))
+        if sha is not None:
+            return sha
+
+        return refs.get('refs/tags/{}'.format(rev))
```

Comment:
These lines are a little hard to parse, especially considering that returning `None` from here is meaningful. How about something like:

```python
branch_ref = 'refs/remotes/origin/{}'.format(rev)
tag_ref = 'refs/tags/{}'.format(rev)

return refs.get(branch_ref) or refs.get(tag_ref)
```
Also, this way, it's not possible for the reader to accidentally conflate the `sha` variable with the `sha` used in the loop above.

There's still an edge case here where a user could have a branch and a tag with the same name, but pointing to different SHAs:

```
$ git show-ref master
df1e0f3fdf3a6a1874447b16f02cb649bc6d4cb5 refs/heads/master
df1e0f3fdf3a6a1874447b16f02cb649bc6d4cb5 refs/remotes/origin/master
37a4c23cbe30d4980bc37b8c9f007ae17433ce20 refs/tags/master
```

But I don't think we can actually disambiguate between them, and it seems to me like the right thing is to prefer the branch (and besides, that's what the current behavior is as well).

Reply:
I would be okay with your suggestion (perhaps with a code comment reminding the reader that `get()` can return None). In the future, I was thinking we might want to return whether a branch or tag was detected for logging purposes, etc, in which case we would need to separate your suggested return statement into more than one line. (I was thinking returning this extra info could be useful if future refactorings result in this function also handling the third case of an "explicit ref," something you previously added support for.)

And yes, you're right about the branch / tag ambiguity. There's a similar preference issue if a branch or tag is given the name of a sha, etc (but that behavior also isn't affected by this PR). One thing I like about the proposed refactoring is that the preference rules are much clearer / more explicit. With the current code, it's much harder to figure out what will happen given various ambiguities and whether those rules are intended (and local branches are also getting mixed in, which shouldn't really be factoring in).